change `_common_parent_group`

ThomasBreuer opened the pull request:
- define and implement the meaning for the group types,
- add tests for the supported situations

resolves #894

ThomasBreuer committed:
change `_common_parent_group`

- define and implement the meaning for the group types,
- add tests for the supported situations

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -216,27 +216,126 @@ function rand_pseudo(G::GAPGroup; radius::Int = 10)
+#
+# - `PermGroup`, `PermGroup`:
+#   The operation is allowed whenever the parents have the same degree,
+#   and the parent is the symmetric group of that degree.
```
Maybe
```suggestion
#   then the new parent is the symmetric group of that degree.
```
or
```suggestion
#   then the parent of the result is the symmetric group of that degree.
```

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -216,27 +216,126 @@ function rand_pseudo(G::GAPGroup; radius::Int = 10)
+#
+# - `PcGroup`, `PcGroup` and
+#   `FPGroup`, `FPGroup`:
+#   The operation is allowed whenever the parents belong to the same `GapObj`
```
I don't understand this. When does a `PcGroup` group "belong" to a `GapObj`?

ThomasBreuer replied:
O.k., let us say "have the same `GapObj`".

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -216,27 +216,126 @@ function rand_pseudo(G::GAPGroup; radius::Int = 10)
+#
+# Note that we cannot guarantee that the groups objects in question are
+# *identical*, only that their `GapObj`s are *equal*:
+# On the GAP side, the same group can be created several times,
```
Maybe add that this applies to perm groups, but not to pc and fp groups?

ThomasBreuer replied:
See the code example about `PcGroup` below.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -216,27 +216,126 @@ function rand_pseudo(G::GAPGroup; radius::Int = 10)
+end
+
+function _common_parent_group(x::PcGroup, y::PcGroup)
+  (x === y || GapObj(x) == GapObj(y)) && return x
```
Does `==` for `PcGroup` do anything besides compare identity?

If so, wouldn't this be sufficient?
```suggestion
  (GapObj(x) === GapObj(y)) && return x
```

ThomasBreuer replied:
The question is what we want to support.
Currently the following works.
```
julia> g = small_group(24, 12);

julia> gg = GAP.Globals.Group(GAP.Globals.GeneratorsOfGroup(GapObj(g)));

julia> newg = PcGroup(gg);

julia> newg == g
true
```
If we want to achieve that two `PcGroup`s are equal only if their `GapObj`s are identical then we have to change `PcGroup`. Likewise for `FPGroup`s. I do not expect difficulties from such a change.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -216,27 +216,126 @@ function rand_pseudo(G::GAPGroup; radius::Int = 10)
+end
+
+function _common_parent_group(x::FPGroup, y::FPGroup)
+  (x === y || GapObj(x) == GapObj(y)) && return x
```
Same questions as for `PCGroup`
```suggestion
  (GapObj(x) === GapObj(y)) && return x
```

ThomasBreuer replied:
(see the comment for `PcGroup`)

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -216,27 +216,126 @@ function rand_pseudo(G::GAPGroup; radius::Int = 10)
+function _common_parent_group(x::SubPcGroup, y::SubPcGroup)
+  x === y && return x
+  @req x.full_group == y.full_group "the groups belong to different full groups"
+  return x.full_group
```
So this is not type stable then, it sometimes gives a `SubPcGroup` and sometimes a `PcGroup`. Hrm, that's annoying, but I see no immediate way to resolve it, unless we were to make the "full group" deliberately a `SubPcGroup` that is an index 1 subgroup of the `PcGroup`...

ThomasBreuer replied:
Taking that index 1 subgroups in `PcGroup`s occur already in many places because of these type conventions (the first member in `derived_series(G)` etc.), we can return always a `SubPcGroup`.
Currently the `SubPcGroup` stores the full `PcGroup`, thus fetching this `.full_group` is cheap, compared to wrapping a new `SubPcGroup`. Perhaps the `SubPcGroup` should store an index 1 subgroup instead?

ThomasBreuer replied:
Hmm, if we want that a `SubPcGroup` stores a `SubPcGroup` as its `full_group` then we have the logical problem that the `SubPcGroup` constructor takes a `GapObj` and calls `SubPcGroup` for its full group, an infinite recursion ...

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -216,27 +216,126 @@ function rand_pseudo(G::GAPGroup; radius::Int = 10)
+end
+
+function _common_parent_group(x::SubFPGroup, y::SubFPGroup)
+  x === y && return x
```
Same as for `PcGroup` :-/

fingolfin commented on src/Groups/perm.jl:
```diff
@@ -286,7 +286,9 @@ function cperm()
-  return prod([PermGroupElem(symmetric_group(maximum(y)), GAPWrap.CycleFromList(GAP.Obj([Int(k) for k in y]))) for y in [L1, L...]])
+  cycles = [L1, L...]
+  n = maximum(map(maximum, cycles))
+  return prod([PermGroupElem(symmetric_group(n), GAPWrap.CycleFromList(GAP.Obj([Int(k) for k in y]))) for y in cycles])
```
Argh, this reminds me I really need to submit a PR for my rewrite of `cperm` and friends.


Also, this method is still inefficient because it creates a fresh `symmetric_group` each time. Either we cache those (see issue #129) or perhaps we should just disallow this without an explicit parent group (well, not immediately to stay compatible with 1.0, but at least we could deprecate this). Thoughts, @fieker ?

fingolfin reviewed (commented):
Thank you! Some quick thoughts

ThomasBreuer committed:
extend the changes to missing group types

- support `_common_parent_group` for `DirectProductGroup`,
  `SemidirectProductGroup`, `WreathProductGroup`
- add `x^y` for group elements (of not nec. equal types)

ThomasBreuer committed:
address comments

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -224,50 +224,64 @@ end
+#   (in the sense of `===´),
+#   then the `full_group` of the `SubPcGroup` (`SubFPGroup`) is taken
+#   as the parent of the result.
+#   Note that we take a group of type `SubPcGroup` (`SubFpGroup`).
```
Perhaps we should explain why (i.e. type stability, e.g. when multiplying elements with identical parent)
